packages/deployer/src/utils/compiler.ts

```diff
@@ -1,12 +1,107 @@
 import * as path from 'path';
 import * as solc from 'solc';
 
 import { constants } from './constants';
 import { fsWrapper } from './fs_wrapper';
-import { ContractArtifact, ContractSources } from './types';
+import { ContractArtifact, ContractSources, FunctionNameToSeenCount } from './types';
 
+/**
+ * Constructs a system-wide unique identifier for a source file.
+ * @param directoryNamespace Namespace of the source file's root contract directory.
+ * @param sourceFilePath Path to a source file, relative to contractBaseDir.
+ * @return sourceFileId A system-wide unique identifier for the source file.
+ */
+export function constructUniqueSourceFileId(directoryNamespace: string, sourceFilePath: string): string {
+    let namespacePrefix = '';
+    if (!_.isEmpty(directoryNamespace)) {
+        namespacePrefix = `/${directoryNamespace}`;
+    }
+    const sourceFilePathNoLeadingSlash = sourceFilePath.replace(/^\/+/g, '');
+    const sourceFileId = `${namespacePrefix}/${sourceFilePathNoLeadingSlash}`;
+    return sourceFileId;
+}
+/**
+ * Constructs a system-wide unique identifier for a dependency file.
+ * @param dependencyFilePath Path from a sourceFile to a dependency.
+ * @param  contractBaseDir Base contracts directory of search tree.
+ * @return sourceFileId A system-wide unique identifier for the source file.
+ */
+export function constructDependencyFileId(dependencyFilePath: string, sourceFilePath: string): string {
+    if (_.startsWith(dependencyFilePath, '/')) {
+        // Path of the form /namespace/path/to/dependency.sol
+        return dependencyFilePath;
+    } else {
+        // Dependency is relative to the source file: ./dependency.sol, ../../some/path/dependency.sol, etc.
+        // Join the two paths to construct a valid source file id: /namespace/path/to/dependency.sol
+        return path.join(path.dirname(sourceFilePath), dependencyFilePath);
+    }
+}
+/**
+ * Constructs a system-wide unique identifier for a contract.
+ * @param directoryNamespace Namespace of the source file's root contract directory.
+ * @param sourceFilePath Path to a source file, relative to contractBaseDir.
+ * @return sourceFileId A system-wide unique identifier for contract.
+ */
+export function constructContractId(directoryNamespace: string, sourceFilePath: string): string {
+    let namespacePrefix = '';
+    if (!_.isEmpty(directoryNamespace)) {
+        namespacePrefix = `${directoryNamespace}:`;
+    }
+    const sourceFileName = path.basename(sourceFilePath, constants.SOLIDITY_FILE_EXTENSION);
+    const contractId = `${namespacePrefix}${sourceFileName}`;
+    return contractId;
+}
+/**
+ *  There is no function overloading in typescript, so we must change the Typescript ABI interface.
+ * Overloaded function names are incremented as follows: functionName, functionName_2, functioname_3, ...
+ * If functionName_N already exists then compilation will fail.
+ */
+export function renameOverloadedFunctionNames(contractAbi: ContractAbi) {
```

Comment:
we might want to return a new `ContractAbi` instance instead of mutating the input, if we choose to stick with this option we might want to document it in the comments above or make it part of the function name